### ISSUE TITLE
dask.array.fft.fftn wrapping mkl_fft segfaults under various conditions

### ISSUE THREAD

**Issue by Jerror** [dask/dask#3563]
If you're lucky, that is -- sometimes it fails silently. This bug affects everyone using an up-to-date Anaconda distribution.

(r)fftn and its convenience wrappers (eg., fft2, rfft2) work as expected when axes=None or when the entire input array is one chunk but segfault when the transform axes are specified and any other axes are not in a single chunk, eg.:
`import dask.array as da`
These succeed and give the expected output:
`da.fft.fftn(da.ones((50,50), chunks=(-1,-1))).compute()`
`da.fft.fftn(da.ones((2,50,50), chunks=(-1,-1,-1)), axes=(1,2)).compute()`.
`da.fft.fftn(da.ones((50,50,2), chunks=(-1,-1,-1)), axes=(0,1)).compute()`.
These segfault:
`da.fft.fftn(da.ones((50,50,2), chunks=(-1,-1,1)), axes=(0,1)).compute()`
`da.fft.fftn(da.ones((2,50,50), chunks=(1,-1,-1)), axes=(1,2)).compute()`.
These gives the expected output (`ValueError: Dask array only supports taking an FFT along an axis that has a single chunk`):
`da.fft.fftn(da.ones((50,50,2), chunks=(-1,-1,1))).compute()`
`da.fft.fftn(da.ones((2,50,50), chunks=(1,-1,-1))).compute()`.
This does not depend on the data, only the dimensions; the same behaviour occurs with `da.random.normal` instead of `da.ones`, and with rfftn as well as fftn.

Strangely, stacking rfftn transformed arrays can occasionally cause a segfault for random input data.
This gives the correct result:
`l = [da.fft.rfftn(da.ones((50,50), chunks=(-1,-1))) for _ in range(2)]`
`da.stack(l).compute()`
This randomly segfaults:
`l = [da.fft.rfftn(da.random.normal(size=(50,50), chunks=(-1,-1))) for _ in range(2)]`
`da.stack(l).compute()`
while fftn works in both cases.

I believe this is strictly a dask + mkl_fft integration bug, not a flaw in mkl_fft. The latter likely does not treat memory as dask expects. From https://github.com/IntelPython/mkl_fft: 
"Since MKL FFT supports performing discrete Fourier transforms over non-contiguously laid out arrays, MKL can be directly used on any well-behaved floating point array with no internal overlaps for both in-place and not in-place transforms of arrays in single and double floating point precision.
This eliminates the need to copy input array contiguously into an intermediate buffer."
This optimization is the strongest benefit of mkl_fft and, I speculate, the point where dask integration breaks down. Wrapping of mkl_fft._numpy_fft.(r)fftn may have to be handled specially.

**Comment by martindurant** [comment]
To check whether this is a thread-stomping problem, you can try `export MKL_NUM_THREADS=1` for your process, or running dask with processes instead of threads (with the multiprocessing or distributed scheduler).

**Comment by Jerror** [comment]
Thanks! Changing MKL_NUM_THREADS had no effect, but all errors disappear when using the multiprocessing scheduler.

**Comment by Jerror** [comment]
Are there any other possible workarounds (or hope for a fix)? Distributed memory parallelism is intolerably slow for my task; I'm nearly memory-bound as it is.

**Comment by jakirkham** [comment]
Would be good to have some more information. For example what OS are we talking about? Some system information (e.g. processor, memory, etc.). What packages from `conda`/`pip` are needed to produce the MRE.

Other questions about the MRE: Are you able to verify this does not occur with MKL using NumPy Arrays? Is the bug reproducible with basic Python threads (i.e. no Dask)? Have you explored [thread affinity]( https://software.intel.com/en-us/node/522691 ) (i.e. pinning thread to cores)? Have you reached out on the Intel support forum to get any advice?

**Comment by Jerror** [comment]
Good questions; I'll look into them as I find time. I haven't brought this to the Intel forum yet.

This occurs on every machine I've tested: my personal computers (up-to-date arch linux) and the Compute Canada National systems (clusters running kernel 3.10 centOS 7). All were running the stock anaconda distribution ("conda install anaconda").

I've only encountered the error using dask.array and the threaded scheduler, so far.

Right now my workaround is to set up a conda environment with the nomkl package, pin nomkl to 3.0.0, and install anaconda -- it builds off openblas, skips mkl_fft, and all is well.
